### DISCUSSION TITLE
parse odata / graphql filters to preset filters

### DISCUSSION THREAD

**Discussion by zewa666** [ghiscoding/slickgrid-universal#1672] [Q&A]
Hey ho, I wonder whether any of you @ghiscoding and @jr01 ever had the requirement to actually parse an odata or graphql filter query so that you can preset your default grids filters.

our use case is to open a grid, prefiltered with one  or multiple ids from a link in a powerbi report.

I made a somewhat hacky approach by regexing the querystring filter for specific col eq xyz cases but wondered whether you guys had different approaches

**Comment by ghiscoding** [comment]
I'm not exactly sure what you mean by parsing the filter query but the [Example 10 - GraphQL](https://ghiscoding.github.io/slickgrid-universal/#/example10) already shows multiple filters as presets which is what I've done in the past for our project. I mean don't you already know the Ids that you need to preset before opening the grid? Filtering the query string should be last resource

**Reply by zewa666** [reply]
the filtering can be on any column with either one or more ids and will be passed via querystring to my app. since we're odata based and those are also well suited for serialization it is what I'd get.

the grid in our case is a dynamic grid rendering app that might get embedded (iframe or nested app) from multiple other dashboards/apps hence again odata is a well suited common language

**Comment by jr01** [comment]
Here's what we do in our app. We have some components that route to other components (using anchors, buttons, etc) using a `filters` url query parameter that contains a json'd `CurrentFilters[]`.

Easier to explain with an example/code. Imagine a list of accounts and you want to zoom into the transactions of a selected account.

accounts.component.ts:
```ts
public transactionFilters: CurrentFilter[] =  [{ columnId: 'account/code', operator: '=', searchTerms: [selectedAccountCode] }]
```

accounts.component.html:
```html
<a    [routerLink]="['/transactions']"        [queryParams]="{ filters: transactionFilters | json }"
```

Clicking the link you get: `https://domain/transactions?filters=%5B%0A%20%20%7B%0A%20%20%20%20%22columnId%22:%20%22account%2Fcode%22,%0A%20%20%20%20%22operator%22:%20%22%3D%22,%0A%20%20%20%20%22searchTerms%22:%20%5B%0A%20%20%20%20%20%20%220110%22%0A%20%20%20%20%5D%0A%20%20%7D%0A%5D`

(url decode the filters query parameter to get the json)

transactions.component.ts
```
ngOninit() {
  this._activatedRoute.queryParamMap.subscribe(params => {
      const filters = params.get('filters');
      if (filters !== null) {
        const json = JSON.parse(filters) as CurrentFilter[];
        this._filters = json;
      }
    });
```

About the oData part in your question. It's of course up to you to decide what API your grid component wants to expose to the outside world. Personally, I wouldn't use oData in that interface for a few reasons:
1. oData is intended for REST and that is not what your grid component exposes to its users, or in other words if I GET https://yourdomain/yourgrid?putheodatastuffhere I imagine that I will receive a HTML with the slickgrid stuff in it and not a REST response
2. oData has so many options that don't translate to the Angular slickgrid grid's presets.

**Reply by zewa666** [reply]
the thing is the feature is the first step to support sharing links for users so you're right that while we support only a subset of odata, its explicitely nit a rest api but similar to Jira queryParams
